asmcomp/deadcode.ml

```diff
@@ -30,48 +35,110 @@ let rec deadcode i =
-       Reg.add_set_array i.live arg)
+      let dc = Array.map deadcode cases in
+      let cases' = Array.map (fun c -> c.i) dc in
+      let s = deadcode i.next in
+      { i = {i with desc = Iswitch(index, cases'); next = s.i};
+        regs = Reg.add_set_array i.live arg;
+        exits = Array.fold_left
+                  (fun acc c -> Int.Set.union acc c.exits) s.exits dc;
+      }
   | Icatch(rec_flag, handlers, body) ->
-      let (body', _) = deadcode body in
-      let handlers' =
-        List.map (fun (nfail, handler) ->
-            let (handler', _) = deadcode handler in
-            nfail, handler')
-          handlers
+    let body' = deadcode body in
+    let s = deadcode i.next in
+    let handlers' = Int.Map.map deadcode (Int.Map.of_list handlers) in
+    (* Previous passes guarantee that indexes of handlers are unique
+       across the entire function and Iexit instructions refer
+       to the correctly scoped handlers.
+       We do not rely on it here, for safety. *)
+    let rec add_live nfail (live_exits, used_handlers) =
+      if Int.Set.mem nfail live_exits then
+        (live_exits, used_handlers)
+      else
+        let live_exits = Int.Set.add nfail live_exits in
+        match Int.Map.find_opt nfail handlers' with
+        | None -> (live_exits, used_handlers)
+        | Some handler ->
+          let used_handlers = (nfail, handler) :: used_handlers in
+          match rec_flag with
+          | Cmm.Nonrecursive -> (live_exits, used_handlers)
+          | Cmm.Recursive ->
+            Int.Set.fold add_live handler.exits (live_exits, used_handlers)
+    in
+    let live_exits, used_handlers =
+      Int.Set.fold add_live body'.exits (Int.Set.empty, [])
+    in
+    (* Remove exits that are going out of scope. *)
+    let used_handler_indexes = Int.Set.of_list (List.map fst used_handlers) in
+    let live_exits = Int.Set.diff live_exits used_handler_indexes in
+    (* For non-recursive catch, live exits referenced in handlers are free. *)
+    let live_exits =
+      match rec_flag with
+      | Cmm.Recursive -> live_exits
+      | Cmm.Nonrecursive ->
+        List.fold_left (fun exits (_,h) -> Int.Set.union h.exits exits)
+          live_exits
+          used_handlers
+    in
+    let rec patch_next body next_final =
+      let next =
+        match body.next.desc with
+        | Iend -> next_final
+        | _ -> patch_next body.next next_final
       in
-      let (s, _) = deadcode i.next in
-      ({i with desc = Icatch(rec_flag, handlers', body'); next = s}, i.live)
-  | Iexit _nfail ->
-      (i, i.live)
+      { body with next; }
+    in
+    let i =
+      match used_handlers with
+      | [] -> (* Simplify catch without handlers *)
+        patch_next body'.i s.i
+      | _ ->
+        let handlers = List.map (fun (n,h) -> (n,h.i)) used_handlers in
+        { i with desc = Icatch(rec_flag, handlers, body'.i); next = s.i }
+    in
+    { i;
+      regs = i.live;
```

Comment:
I think this needs to be `regs = Reg.add_set_array i.live arg` since `i` could now be an instruction with arguments.

Reply:
Yes, but I think we can just use `regs = body'.regs` in this case.